## Final Predictions

In [22]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.linear_model import LinearRegression

In [2]:
df_new2 = pd.read_csv('Finaldata', index_col=0)

### Train Test Split

We will now conduct a train-test-split on our dataset with a train size of 80% and a test size of 20%. 

In [3]:
X = df_new2.drop('price', axis=1)
y = df_new2['price']

In [4]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [17]:
# train model
linreg = LinearRegression()
linreg.fit(X_train, y_train)

# make predictions
y_hat_train = linreg.predict(X_train)
y_hat_test = linreg.predict(X_test)


# calculate MSE 
train_mse = mean_squared_error(y_train, y_hat_train)
test_mse = mean_squared_error(y_test, y_hat_test)
print('Train Mean Squarred Error:', train_mse)
print('Test Mean Squarred Error:', test_mse)



# calculate RMSE 
train_rmse = np.sqrt(mean_squared_error(y_train, y_hat_train))
test_rmse = np.sqrt(mean_squared_error(y_test, y_hat_test))
print('Train Root Mean Squarred Error:', train_rmse)
print('Test Root Mean Squarred Error:', test_rmse)


Train Mean Squarred Error: 5699200634.298144
Test Mean Squarred Error: 5798429846.170465
Train Root Mean Squarred Error: 75493.0502384037
Test Root Mean Squarred Error: 76147.42179595094


The RMSE for our train level is lower than the RMSE for our test set, which makes sense since we trained the model on our train data. Our RMSE is easier to interpret; firstly, our values are quite low. More precisely, predictions for our train data are typically off by \\$75,493 while our test data predictions are off by \\$76,147. 

### Run Cross Validation

In [18]:
from sklearn.metrics import make_scorer
from sklearn.model_selection import cross_val_score

In [20]:
# K-Fold Cross-Validation with 30 folds
mse = make_scorer(mean_squared_error)
cv_30_results = cross_val_score(linreg, X, y, cv=30, scoring=mse)
cv_30_results.mean()

5744596576.790842

Our 30-fold cross validation returns a value that is close to the train and test MSE values that we ran previously.

In [24]:
# Check R2 score for the train-test split
print('Train Data', r2_score(y_train, linreg.predict(X_train)))
print('Test Data', r2_score(y_test, linreg.predict(X_test)))

print('Our regression can predict home price with approximately 70% accuracy.')

Train Data 0.7002809159095658
Test Data 0.7019363713349078
Our regression can predict home price with approximately 70% accuracy.


### Conclusion

Our model is able to predict housing price with approximately 70% accuracy, which is a significant improvement from our base model. The average mean squared error of a K-fold cross-validation of 30 folds is similar to the MSE values of our train and test sets. Additionally, our model has quite low RMSE values for both our train and our test sets; meaning that on average our model misses our values by \\$75,493 for our train set and \\$76,147 for our test set. Our test and train set RMSEs are quite close together, so we can assume that our model is not overfit nor underfit, and that our regression does a fair job of estimating our testing data. 

Additionally, our model has made several assumptions. The residuals are normally distributed, features are not correlated with one another and therefore our model has limited multicollinearity, the residuals follow homoscedasticity, and, since we dropped all p-values that are less than 0.05, our remaining features are linearly related.

### Business Recommendation

We now will revisit our business problem that was explained at the beginning of our project: assisting a housing development company determine which areas of King County have a positive relationship with price, as well as which features of these homes result in higher priced homes. 

Based on our results, we see that homes in the following areas are positively related to price: Bellevue, Bothell, Fall City, Issaquah, Kirkland, Mercer Island, North Bend, Redmond and Vashon. Our client should consider building their new developement in one of the listed areas, as having a home in these areas is bound to have an increase on the price and value of the home.

Additionally, certain features of homes will also positively impact their price: a nice view of the property, good condition of the homes, and a high square footage of the home's living space, especially for homes with a larger number of floors and bathrooms. Therefore, in order to increase the price of the homes in this new developemnt, the housing company should consider including these features in the models of their new builds. 

### Additional Work

Future work could be done in order to strengthen the predictability of our model; we would consider investigating the following:

- Polynomials: we could introduce polynomials to the model to see if these values better represent the relationship between certain features and price
- Engineer new features, look at ratios between features 
- Stepwise selection to determine if any new features, polynomials, additional interactions or dummy variables in order to see which elements improve our model